src/main/java/nova/internal/core/util/TopologicalSort.java

```diff
@@ -0,0 +1,164 @@
+package nova.internal.core.util;
+
+import nova.core.util.exception.NovaException;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.Set;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+public class TopologicalSort {
+	/**
+	 * Sort the input graph into a topologically sorted list
+	 *
+	 * Uses the reverse depth first search as outlined in ...
+	 * @param graph
+	 * @return The sorted list.
+	 */
+	public static <T> List<T> topologicalSort(DirectedGraph<T> graph) {
+		DirectedGraph<T> rGraph = reverse(graph);
+		List<T> sortedResult = new ArrayList<>();
+		Set<T> visitedNodes = new HashSet<>();
+		// A list of "fully explored" nodes. Leftovers in here indicate cycles in the graph
+		Set<T> expandedNodes = new HashSet<>();
+
+		for (T node : rGraph) {
+			explore(node, rGraph, sortedResult, visitedNodes, expandedNodes);
+		}
+
+		return sortedResult;
+	}
+
+	public static <T> DirectedGraph<T> reverse(DirectedGraph<T> graph) {
+		DirectedGraph<T> result = new DirectedGraph<>();
+
+		for (T node : graph) {
+			result.addNode(node);
+		}
+
+		for (T from : graph) {
+			for (T to : graph.edgesFrom(from)) {
+				result.addEdge(to, from);
+			}
+		}
+
+		return result;
+	}
+
+	public static <T> void explore(T node, DirectedGraph<T> graph, List<T> sortedResult, Set<T> visitedNodes, Set<T> expandedNodes) {
+		// Have we been here before?
+		if (visitedNodes.contains(node)) {
+			// And have completed this node before
+			if (expandedNodes.contains(node)) {
+				// Then we're fine
+				return;
+			}
+
+			throw new NovaException("There was a cycle detected in the input graph, sorting is not possible", node) {
```

Comment:
We shouldn't throw plain NovaEvents really, and do we want the whole game to crash with a dependency cycle? I would check when adding the event so you get a proper stack trace, and isn't adding events less common, so less of a performance issue.